[Taufiq Bin Abdul Rahman] iP

tau-bar opened the pull request:
# Duke
_Duke_ is:
* a task manager
* ~hard~ **easy** to use
* cool
  * awesome

1. download from [here](https://github.com/tau-bar/ip)
1. use it

Kanye West said:

> **use it**!

`System.out.println("Duke is so cool!")`

I have:
- [x] made it cool 💯 

Duke is | Duke is not
------------ | -------------
A cool task manager | Anything else
Easy to use | Hard to use

tau-bar committed:
Added ability to echo commands given to chatbot.

tau-bar committed:
Added Task and TaskList classes, ability to add and list tasks.

tau-bar committed:
Added ability to mark tasks as done.

tau-bar committed:
Changed the way tasks are listed by using toString() method.

tau-bar committed:
Created new classes to support multiple task types.

tau-bar committed:
Added support for tracking todos, deadlines, events.

tau-bar committed:
Added javadocs for Todo, Event and Deadline.

tau-bar committed:
Use the input/output redirection technique to semi-automate the testing of Duke.

tau-bar committed:
Teach Duke to deal with errors such as incorrect inputs entered by the user.

tau-bar committed:
Added error for out of range done command.

tau-bar committed:
Changed tasks to be stored in ArrayList.

tau-bar committed:
Added support for deleting tasks from the list.

tau-bar committed:
Added use of enums for TaskType.

tau-bar committed:
Changed from if-else to using switch block for Duke commands.

tau-bar committed:
Put each command into separate functions, and made tasks a static field in Duke class.

tau-bar committed:
Add ability to save tasks in hard disk and read from hard disk

tau-bar committed:
Teach Duke how to understand dates and times

tau-bar committed:
Merge tag 'Level-7'

tau-bar committed:
Merge Level-8 to master

tau-bar committed:
Make code more OOP

tau-bar committed:
Organize classes into duke package

tau-bar committed:
Add JUnit tests to test behavior of code

tau-bar committed:
Package the app as an executable JAR file so that it can be distributed easily

tau-bar committed:
Add JavaDoc comments to code

tau-bar committed:
Tweak the code to comply with a coding standard

tau-bar committed:
Add feature allowing users to find a task by searching for a keyword

tau-bar committed:
Merge branch 'branch-Level-9'

tau-bar committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/DukeUI.java
#	src/main/java/duke/TaskList.java

tau-bar committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Deadline.java
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Event.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java

tau-bar committed:
Update README.md

tau-bar commented:
# Duke
_Duke_ is:
* a task manager
* ~hard~ **easy** to use
* cool
  * awesome

1. download from [here](https://github.com/tau-bar/ip)
1. use it

Kanye West said:

> **use it**!

`System.out.println("Duke is so cool!")`

I have:
- [x] made it cool 💯 

Duke is | Duke is not
------------ | -------------
A cool task manager | Anything else
Easy to use | Hard to use

tau-bar committed:
Fix failed text test

tau-bar committed:
Merge branch 'master' of https://github.com/tau-bar/ip

tau-bar committed:
Add gradle support

tau-bar committed:
Merge branch 'add-gradle-support'

tau-bar committed:
Add checkstyle support

tau-bar committed:
Merge branch 'add-checkstyle'

joshenx commented on .gitignore:
```diff
@@ -13,5 +13,14 @@ src/main/resources/docs/
+# Ignore .class files
+*.class
+
+#Ignore artifact directory
```
```suggestion
# Ignore artifact directory
```

joshenx replied:
```suggestion
# Ignore artifact directory
```

joshenx commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    /**
+     * Method to delete task.
```
I think that Javadoc method header comments tend to describe what they do without starting as "Method to...". Perhaps a better comment is "Deletes a task from the list."? I noticed similar header comments as well.

joshenx commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,204 @@
+        String taskDesc = this.parser.findTaskDescription();
+        String aOrAn = firstCommand.equals("event") ? "an" : "a";
+        if (taskDesc.equals("")) {
+            throw new DukeException("☹ OOPS!!! The description of " + aOrAn + " " + firstCommand + " cannot be empty.");
```
Should these error messages be abstracted into final Strings so that they can be easily modified when the errors are thrown? For example, you could consider having methods in the ```DukeException``` class to return a ```DukeException``` instance with a specific type of error message:

```java
public class DukeException extends Exception {
    public DukeException(String message) {
      super(message);
    }
    //...
    public static final String EMPTY_DESC_ERROR = "Empty description error!";
    //...
    public static DukeException emptyDescError() {
      return new DukeException(EMPTY_DESC_ERROR);
    }
}
```
Then, throw the exception with
```throw DukeException.emptyDescError();```

joshenx commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,204 @@
+
+    /**
+     * Method for duke.Duke to mark a task done.
+     * @throws DukeException Thrown when user gives an index out of range
```
```suggestion
     * Method for duke.Duke to mark a task done.
     *
     * @throws DukeException Thrown when user gives an index out of range
```

joshenx commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    /**
+     * Main method for Duke.
+     * @param args Optional arguments for CLI.
```
```suggestion
     * Main method for Duke.
     *
     * @param args Optional arguments for CLI.
```

I noticed this in other javadoc header comments as well! There should be an empty line between description and parameter section, as per the [Coding Standard](https://se-education.org/guides/conventions/java/intermediate.html).

joshenx reviewed (approved):
LGTM, though I did noticed some minor coding standard violations. I'm impressed at how indepth your Javadoc comments are, and how you managed to abstract many functions of the program into methods. Do review my suggestion of further abstracting DukeException within the class itself, and let me know if you have any clarifications. Good job!

tau-bar committed:
Add GUI using JavaFX

tau-bar committed:
Merge branch 'branch-Level-10'

tau-bar committed:
Update .gitignore

Co-authored-by: Joshen Lim <77193243+joshenx@users.noreply.github.com>

tau-bar committed:
Update src/main/java/duke/Duke.java

Co-authored-by: Joshen Lim <77193243+joshenx@users.noreply.github.com>

tau-bar committed:
Update src/main/java/duke/Duke.java

Co-authored-by: Joshen Lim <77193243+joshenx@users.noreply.github.com>

tau-bar committed:
Update java docs to follow coding standard

tau-bar committed:
Merge branch 'master' of https://github.com/tau-bar/ip

yulonglim commented on src/main/java/duke/Duke.java:
```diff
@@ -42,70 +37,88 @@ public Duke(String filePath) {
+    private String respondToFirstCommand(String firstCommand) {
+        String response;
+        switch (firstCommand) {
+        case "bye":
```
suggest making firstCommand ignore case or to set to lowercase to reduce user input error?

yulonglim commented on src/main/java/duke/DukeUI.java:
```diff
@@ -6,45 +6,49 @@
      */
-    public void greetUser() {
+    public String greetUser() {
         String logo = " ____        _        \n"
```
if logo is still used, the name should be all capitalized as it is a constant.

yulonglim commented on src/main/java/duke/Duke.java:
```diff
@@ -117,7 +130,7 @@ private void writeDataToDuke() {
+    public String addTask() throws DukeException {
         String firstCommand = this.parser.getFirstCommand();
         String date = this.parser.findDateInCommand();
         String taskDesc = this.parser.findTaskDescription();
```
maybe write the whole name of variable?

yulonglim commented on src/main/java/duke/Duke.java:
```diff
@@ -134,40 +147,40 @@ public void addTask() throws DukeException {
      */
-    public void confirmAdditionOfTask() {
+    public String confirmAdditionOfTask() {
         int tasksLength = this.tasks.getTasksLength();
```
tasksLength may be confusing, might refer to task duration,etc ? Suggest changing variable name to taskListLength

yulonglim reviewed (commented):
Code is generally well written, although some method name are long, they do not violate java coding standards!

tau-bar committed:
Change commands to be case insensitive

Commands are case sensitive

A case insensitive command is more user-friendly as users cannot be expected to remember the exact case of the keywords

Let's
* update the tests to check if case insensitive commands work
* refactor the rest of the code to improve code quality

tau-bar committed:
Revert "Change commands to be case insensitive"

This reverts commit fa3da5534c2b9b49e1d30418f748b47ba79d100f.

tau-bar committed:
Use assert feature

Currently, there are no safety measures in the functional code.

Add assertion statements to document important assumptions that should hold at various points in the code

Assertions add additional safety measures inside the functional code

tau-bar committed:
Improve code quality

There are some violations of code quality standards

Edit code to adhere to code quality guidelines

Improve code quality for easier readability and understanding

tau-bar committed:
Merge pull request #3 from tau-bar/branch-A-Assertions

Use assert feature

tau-bar committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Duke.java
#	src/main/java/duke/Parser.java

tau-bar committed:
Merge pull request #4 from tau-bar/branch-A-CodeQuality

Improve code quality

tau-bar committed:
Add ability to see schedule and specify date that want to see schedule

Some tasks have dates, but no way to see schedule

Add ability to see schedule allows user to see what tasks have dates and what tasks are on a certain date.

As step to better functionality add ability to see schedule or schedule on certain day

tau-bar committed:
Merge branch 'branch-B-ViewSchedules'

tau-bar committed:
Improve code quality

tau-bar committed:
Fix bug in GUI and resolve checkstyle issues

tau-bar committed:
Improve GUI to highlight errors with different background colour

tau-bar committed:
Rename duke to taubot to finalise features

tau-bar committed:
Change personality of bot, fixed schedule function if no date passed in

tau-bar committed:
Add User Guide with features and quick start

tau-bar committed:
Add command summary to User Guide

tau-bar committed:
Add acknowldegment section in README

tau-bar committed:
Set theme jekyll-theme-cayman

tau-bar committed:
Update README.md

tau-bar committed:
Update README.md

tau-bar committed:
Update README.md

tau-bar committed:
Add UI.png

tau-bar committed:
Rename UI.png to Ui.png

tau-bar committed:
Fix bug where no tasks matching keyword

tau-bar committed:
Move Ui.png into docs folder

tau-bar committed:
Fix bug where new file not created if database is missing

tau-bar committed:
Move Ui.png into docs folder

tau-bar committed:
Update build.gradle to launch GUI app

tau-bar committed:
Add anchor links

tau-bar committed:
Update README.md

tau-bar committed:
Update README.md

tau-bar committed:
Update README.md

tau-bar committed:
Update README.md

tau-bar committed:
Update README.md

tau-bar committed:
Update README.md

tau-bar committed:
Change visibility of some methods

tau-bar committed:
Merge branch 'master' of https://github.com/tau-bar/ip
